neuralpp/symbolic/expression.py

```diff
@@ -0,0 +1,151 @@
+from typing import Any, List
+from abc import ABC, abstractmethod
+
+
+class Expression(ABC):
+    @abstractmethod
+    # 'Expression': forward reference for recursive type reference https://peps.python.org/pep-0484/#forward-references
+    def subexpression(self) -> List['Expression']:
+        """
+        Returns a list of subexpression.
+        E.g.,
+        subexpression(f(x,y)) = [f,x,y]
+        subexpression(add(a,1)) = [add,a,1]
+        """
+        pass
+
+    @abstractmethod
+    def set(self, i: int, new_expression: 'Expression') -> 'Expression':
+        """
+        Set i-th subexpression to new_expression. Count from 0.
+        E.g.,
+        f(x,y).set(1,z) = [f,z,y].
+        If it's out of the scope, return error.
+        """
+        pass
+
+    @abstractmethod
+    def replace(self, from_expression: 'Expression', to_expression: 'Expression') -> 'Expression':
+        """
+        Every expression is immutable so replace() returns either self or a new Expression.
+        No in-place modification should be made.
+        If from_expression is not in the expression, returns self.
+        """
+        pass
+
+    def contains(self, target: 'Expression') -> bool:
```

Comment:
Yeah maybe that makes more sense. In that case `Expression` can be purely abstract.

Reply:
per @rodrigodesalvobraz's suggestion of a new `contains()` semantics, now it might be better to have `contains()` in `Expression`? Since `contains()` of `AtomicExpression` and `FunctionApplication` share the `if self == target: return True` logic, having a `contains()` in `Expression` is more DRY?

Reply:
I think it depends how strongly we want `Expression` to be an abstract class. Not sure if other people have any strong opinions about it. I think it's fair to have the logic be contained in `Expression`.

Reply:
I think either way is ok. The guideline I would follow is: if a method definition is going to apply to *most* Expressions, then it can be placed in Expression as a default. I am not sure if this version here is the latest one, I will go check the latest version and leave a comment there if needed.